File: ags/Bar/Bar.js
import icons from "../icons.js";
import ToggleIconButton from "../widgets/ToggleIconButton.js";
import BatteryWidget from "./widgets/BatteryWidget.js";
import ClockWidget from "./widgets/ClockWidget.js";
import KeyboardLayout from "./widgets/KeyboardLayout.js";
import PowerProfileWidget from "./widgets/PowerProfileWidget.js";
import SysTray from "./widgets/SysTray.js";
import TodoTxtWidget from "./widgets/TodoTxtWidget.js";
import WeatherWidget from "./widgets/WeatherWidget.js";
import Workspaces from "./widgets/Workspaces.js";

function Left() {
  return Widget.Box({
    spacing: 8,
    children: [TodoTxtWidget(), SysTray()],
  });
}

function Center() {
  return Widget.Box({
    spacing: 8,
    children: [WeatherWidget(), ClockWidget(), KeyboardLayout()],
  });
}

function Right(monitor = 0) {
  return Widget.Box({
    hpack: "end",
    spacing: 8,
    children: [
      Workspaces(),
      PowerProfileWidget(),
      BatteryWidget(),
      ToggleIconButton(() => {
        const id = `new-settings-${monitor}`;
        App.toggleWindow(id);
      }, Widget.Icon({ icon: icons.session.shutdown })),
    ],
  });
}

function Bar(monitor = 0) {
  return Widget.Window({
    name: `bar-${monitor}`,
    class_name: "bar",
    monitor,
    anchor: ["top", "left", "right"],
    exclusivity: "exclusive",
    child: Widget.CenterBox({
      start_widget: Left(),
      center_widget: Center(),
      end_widget: Right(monitor),
    }),
  });
}

export default Bar;

File: ignis/statusbar/bar.py
from ignis import utils, widgets
from ignis.services.hyprland import HyprlandService

from .widgets import (
    client_title,
    clock,
    settings_menu,
    tray,
    workspaces,
)

hyprland = HyprlandService.get_default()


def left(monitor_name: str) -> widgets.Box:
    return widgets.Box(child=[workspaces(monitor_name), client_title()], spacing=10)


def center() -> widgets.Box:
    return widgets.Box(
        child=[
            clock(),
        ],
        spacing=10,
    )


def right() -> widgets.Box:
    return widgets.Box(
        child=[
            tray(),
            settings_menu(),
        ],
        spacing=10,
    )


def bar(monitor_id: int = 0) -> widgets.Window:
    monitor_name = utils.get_monitor(monitor_id).get_connector()  # type: ignore
    return widgets.Window(
        namespace=f"ignis_bar_{monitor_id}",
        monitor=monitor_id,
        anchor=["left", "top", "right"],
        exclusivity="exclusive",
        child=widgets.CenterBox(
            css_classes=["bar"],
            start_widget=left(monitor_name),  # type: ignore
            center_widget=center(),
            end_widget=right(),
        ),
    )
